File: minimumrotatedsortedarray.py
#Time Complexity: O(logn)

class Solution:
    def findMin(self, nums: List[int]) -> int:
        low=0
        high=len(nums)-1

        while low<=high:
            #if low<=high means array is sorted
            if nums[low]<=nums[high]:
                #return low directly as min
                return nums[low]
            # if low is not less than high then array is unsorted
            mid=low+(high-low)//2
           
            
            if (mid ==0 or nums[mid]<nums[mid-1]) and nums[mid]<nums[mid+1]:
                return nums[mid]
            #now check again if above case is not true then either of the half is again sorted and eliminate that half 
            elif nums[low]<=nums[mid]:
                # left half is sorted then move low
                low=mid+1
                # else right is sorted then move high 
            else:
                high=mid-1

File: MinimumRotatedSortedArray.java
public class MinimumRotatedSortedArray {
    public int findMin(int[] nums) {
        if(nums.length == 1){
            return nums[0];
        }

        int low =0, high = nums.length-1;
        while(high>=low){
            if(nums[low] <= nums[high]){
                return nums[low];
            }

            int mid = low + (high-low)/2;

            if((mid == nums.length-1 || nums[mid] < nums[mid+1])
                    && (mid==0 || nums[mid] < nums[mid-1])){
                return nums[mid];
            }

            if(nums[low] <= nums[mid]){
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }
}
